### ISSUE TITLE
🚂 Implement Complete Neural Network Training Pipeline

### ISSUE THREAD

**Issue by sakshambedi** [sakshambedi/micrograd#21]
# 🚂 Implement Complete Neural Network Training Pipeline

## 📋 Description

Create a comprehensive training pipeline that integrates all micrograd components to enable end-to-end neural network training. This includes data handling, training loops, validation, checkpointing, and monitoring capabilities.

## 🎯 Acceptance Criteria

- [ ] Implement `Trainer` class for orchestrating training process
- [ ] Add `DataLoader` for batch processing and data iteration
- [ ] Implement training and validation loop management
- [ ] Add model checkpointing and state saving/loading
- [ ] Implement metrics tracking and logging
- [ ] Add early stopping and training callbacks
- [ ] Support for different training strategies (epoch-based, step-based)
- [ ] Implement gradient accumulation for large batch training
- [ ] Add model evaluation utilities
- [ ] Create comprehensive training examples and tutorials

### Current State
- Neural network layers: ✅ Implemented (Issue #08)
- Loss functions: ✅ Implemented (Issue #09)
- Optimizers: ✅ Implemented (Issue #10)
- Autograd system: ✅ Implemented (Issues #01, #03)
- Data pipeline: ❌ Not implemented

### Implementation Requirements

**Trainer Class**
```python
class Trainer:
    """Main training orchestrator."""
    
    def __init__(self, model, optimizer, criterion, device='cpu'):
        self.model = model
        self.optimizer = optimizer
        self.criterion = criterion
        self.device = device
        self.metrics = {}
        self.callbacks = []
        self.history = {'train_loss': [], 'val_loss': []}
    
    def fit(self, train_loader, val_loader=None, epochs=100, 
            validation_freq=1, save_best=True, checkpoint_dir=None):
        """Main training loop."""
        best_val_loss = float('inf')
        
        for epoch in range(epochs):
            # Training phase
            train_loss = self.train_epoch(train_loader)
            self.history['train_loss'].append(train_loss)
            
            # Validation phase
            if val_loader and epoch % validation_freq == 0:
                val_loss = self.validate(val_loader)
                self.history['val_loss'].append(val_loss)
                
                # Save best model
                if save_best and val_loss < best_val_loss:
                    best_val_loss = val_loss
                    self.save_checkpoint(checkpoint_dir, f'best_model.pt')
            
            # Execute callbacks
            for callback in self.callbacks:
                callback.on_epoch_end(epoch, self.history)
    
    def train_epoch(self, train_loader):
        """Single training epoch."""
        self.model.train()
        total_loss = 0.0
        num_batches = 0
        
        for batch_idx, (data, target) in enumerate(train_loader):
            # Forward pass
            output = self.model(data)
            loss = self.criterion(output, target)
            
            # Backward pass
            self.optimizer.zero_grad()
            loss.backward()
            self.optimizer.step()
            
            total_loss += loss.item()
            num_batches += 1
        
        return total_loss / num_batches
    
    def validate(self, val_loader):
        """Validation phase."""
        self.model.eval()
        total_loss = 0.0
        num_batches = 0
        
        for data, target in val_loader:
            output = self.model(data)
            loss = self.criterion(output, target)
            total_loss += loss.item()
            num_batches += 1
        
        return total_loss / num_batches
```

**DataLoader Implementation**
```python
class DataLoader:
    """Simple data loader for batch processing."""
    
    def __init__(self, dataset, batch_size=32, shuffle=True, drop_last=False):
        self.dataset = dataset
        self.batch_size = batch_size
        self.shuffle = shuffle
        self.drop_last = drop_last
        self.indices = list(range(len(dataset)))
    
    def __iter__(self):
        if self.shuffle:
            random.shuffle(self.indices)
        
        for i in range(0, len(self.indices), self.batch_size):
            batch_indices = self.indices[i:i + self.batch_size]
            
            if self.drop_last and len(batch_indices) < self.batch_size:
                continue
            
            batch_data = []
            batch_targets = []
            
            for idx in batch_indices:
                data, target = self.dataset[idx]
                batch_data.append(data)
                batch_targets.append(target)
            
            # Stack into tensors
            batch_data = Tensor.stack(batch_data)
            batch_targets = Tensor.stack(batch_targets)
            
            yield batch_data, batch_targets
    
    def __len__(self):
        if self.drop_last:
            return len(self.dataset) // self.batch_size
        else:
            return (len(self.dataset) + self.batch_size - 1) // self.batch_size
```

**Callback System**
```python
class Callback:
    """Base class for training callbacks."""
    
    def on_epoch_begin(self, epoch, history):
        pass
    
    def on_epoch_end(self, epoch, history):
        pass
    
    def on_batch_begin(self, batch, history):
        pass
    
    def on_batch_end(self, batch, history):
        pass

class EarlyStopping(Callback):
    """Early stopping callback."""
    
    def __init__(self, monitor='val_loss', patience=10, min_delta=0):
        self.monitor = monitor
        self.patience = patience
        self.min_delta = min_delta
        self.best_value = None
        self.wait = 0
        self.stopped_epoch = 0
    
    def on_epoch_end(self, epoch, history):
        current = history[self.monitor][-1]
        
        if self.best_value is None:
            self.best_value = current
        elif current < self.best_value - self.min_delta:
            self.best_value = current
            self.wait = 0
        else:
            self.wait += 1
            if self.wait >= self.patience:
                self.stopped_epoch = epoch
                return True  # Stop training
        
        return False
```

### Model Checkpointing
```python
class ModelCheckpoint:
    """Save model checkpoints during training."""
    
    @staticmethod
    def save(model, optimizer, epoch, loss, filepath):
        """Save model state."""
        checkpoint = {
            'epoch': epoch,
            'model_state': model.state_dict(),
            'optimizer_state': optimizer.state_dict(),
            'loss': loss,
        }
        # Implementation depends on serialization method
        
    @staticmethod
    def load(filepath, model, optimizer=None):
        """Load model state."""
        checkpoint = load_checkpoint(filepath)
        model.load_state_dict(checkpoint['model_state'])
        if optimizer:
            optimizer.load_state_dict(checkpoint['optimizer_state'])
        return checkpoint['epoch'], checkpoint['loss']
```

## 🧪 Testing Requirements

- [ ] Test complete training pipeline with simple datasets
- [ ] Test trainer with different optimizers and loss functions
- [ ] Test DataLoader with various batch sizes and shuffle options
- [ ] Test callback system and early stopping
- [ ] Test model checkpointing and resuming training
- [ ] Test training with validation monitoring
- [ ] Integration tests with all neural network components
- [ ] Memory leak tests during long training runs
- [ ] Performance benchmarks for training speed
- [ ] Test gradient accumulation for large batches

## 📁 Files to Create/Modify

- `grad/training/__init__.py` - Training package
- `grad/training/trainer.py` - Main Trainer class
- `grad/training/dataloader.py` - DataLoader implementation
- `grad/training/callbacks.py` - Callback system
- `grad/training/checkpoint.py` - Model checkpointing utilities
- `grad/training/metrics.py` - Training metrics and monitoring
- `grad/utils/data.py` - Data utilities and dataset classes
- `tests/test_training_pipeline.py` - Complete pipeline tests
- `examples/mnist_training.py` - Complete MNIST training example
- `examples/regression_example.py` - Regression training example

## 🔗 Dependencies

- Requires: Neural network layers (#08)
- Requires: Loss functions (#09)
- Requires: Optimizers (#10)
- Requires: Autograd system (#01, #03)
- Requires: Matrix multiplication (#04)
- Requires: Tensor operations (#06)
- Enables: Production-ready neural network training
- Enables: Research and experimentation workflows

### Simple Classification Training
```python
from grad.training import Trainer, DataLoader
from grad.nn import Sequential, Linear, ReLU
from grad.nn.loss import CrossEntropyLoss
from grad.optim import Adam

# Create model
model = Sequential(
    Linear(784, 128),
    ReLU(),
    Linear(128, 64), 
    ReLU(),
    Linear(64, 10)
)

# Setup training
optimizer = Adam(model.parameters(), lr=0.001)
criterion = CrossEntropyLoss()
trainer = Trainer(model, optimizer, criterion)

# Add callbacks
early_stopping = EarlyStopping(monitor='val_loss', patience=10)
trainer.add_callback(early_stopping)

# Create data loaders
train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=32, shuffle=False)

# Train model
trainer.fit(train_loader, val_loader, epochs=100)

# Evaluate
test_accuracy = trainer.evaluate(test_loader, metric='accuracy')
print(f'Test Accuracy: {test_accuracy:.4f}')
```

### Advanced Training with Callbacks
```python
# Custom callback for learning rate scheduling
class LRSchedulerCallback(Callback):
    def __init__(self, scheduler):
        self.scheduler = scheduler
    
    def on_epoch_end(self, epoch, history):
        self.scheduler.step()

# Setup advanced training
trainer = Trainer(model, optimizer, criterion)
trainer.add_callback(EarlyStopping(patience=15))
trainer.add_callback(LRSchedulerCallback(scheduler))
trainer.add_callback(ModelCheckpoint(save_best=True, save_freq=10))

# Train with gradient accumulation
trainer.fit(
    train_loader, 
    val_loader, 
    epochs=200,
    accumulation_steps=4  # Simulate larger batch size
)
```

### Transfer Learning Example
```python
# Load pre-trained model
model = load_pretrained_model('pretrained_model.pt')

# Freeze early layers
for param in model.layers[:-2].parameters():
    param.requires_grad = False

# Fine-tune on new dataset
optimizer = Adam(model.parameters(), lr=0.0001)
trainer = Trainer(model, optimizer, criterion)
trainer.fit(new_train_loader, new_val_loader, epochs=50)
```

## 🚨 Edge Cases to Handle

- Empty datasets or data loaders
- Model that doesn't converge (loss becomes NaN)
- Memory constraints with large models/batches
- Interrupted training (resume from checkpoint)
- Validation dataset larger than memory
- Different data types in batches
- Gradient explosion during training
- Hardware failures during long training runs

## 🔧 Performance Considerations

- Efficient batch processing and memory management
- Minimize data copying between operations
- Lazy loading for large datasets
- Memory pooling for repeated allocations
- Progress monitoring without performance impact
- Efficient checkpoint serialization
- Consider future multi-GPU support

## 🏷️ Labels

- `enhancement`
- `training-pipeline`
- `machine-learning`
- `high-priority`
- `integration`

## 📈 Priority

**High** - Completes the neural network framework for practical use

## 🔗 Related Issues

- Depends on: All previous issues (#01-#10)
- Enables: Production neural network training
- Enables: Research and experimentation
- Foundation for: Advanced features (distributed training, etc.)
- Foundation for: Real-world applications and benchmarks
